src/wayland/selection/device.rs

```diff
@@ -9,16 +7,89 @@ use wayland_server::protocol::wl_seat::WlSeat;
 use wayland_server::Resource;
 
 use super::data_device::DataDeviceUserData;
+use super::ext_data_control::ExtDataControlDeviceUserData;
+use super::offer::DataControlOffer;
 use super::offer::SelectionOffer;
 use super::primary_selection::PrimaryDeviceUserData;
 use super::private::selection_dispatch;
 use super::wlr_data_control::DataControlDeviceUserData;
 
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub(crate) enum DataDeviceKind {
+    Core,
+    Primary,
+    WlrDataControl,
+    ExtDataControl,
+}
+
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub enum SelectionDevice {
+pub(crate) enum DataControlDevice {
+    Wlr(ZwlrDataControlDeviceV1),
+    Ext(ExtDataControlDeviceV1),
+}
+
+impl DataControlDevice {
+    fn data_offer(&self, offer: &DataControlOffer) {
+        match (self, offer) {
+            (Self::Wlr(obj), DataControlOffer::Wlr(offer)) => obj.data_offer(offer),
+            (Self::Ext(obj), DataControlOffer::Ext(offer)) => obj.data_offer(offer),
+            _ => unreachable!(),
+        }
+    }
+
+    fn selection(&self, offer: Option<&DataControlOffer>) {
+        match (self, offer) {
+            (Self::Wlr(obj), Some(DataControlOffer::Wlr(offer))) => obj.selection(Some(offer)),
+            (Self::Ext(obj), Some(DataControlOffer::Ext(offer))) => obj.selection(Some(offer)),
+            (Self::Wlr(obj), None) => obj.selection(None),
+            (Self::Ext(obj), None) => obj.selection(None),
+            _ => unreachable!(),
+        }
+    }
+
+    fn primary_selection(&self, offer: Option<&DataControlOffer>) {
+        match (self, offer) {
+            (Self::Wlr(obj), Some(DataControlOffer::Wlr(offer))) => obj.primary_selection(Some(offer)),
+            (Self::Ext(obj), Some(DataControlOffer::Ext(offer))) => obj.primary_selection(Some(offer)),
+            (Self::Wlr(obj), None) => obj.primary_selection(None),
+            (Self::Ext(obj), None) => obj.primary_selection(None),
+            _ => unreachable!(),
+        }
+    }
+
+    fn wl_seat(&self) -> WlSeat {
+        match self {
+            Self::Wlr(device) => {
+                let data: &DataControlDeviceUserData = device.data().unwrap();
+                data.wl_seat.clone()
+            }
+            Self::Ext(device) => {
+                let data: &ExtDataControlDeviceUserData = device.data().unwrap();
+                data.wl_seat.clone()
+            }
+        }
+    }
+
+    fn id(&self) -> ObjectId {
+        match self {
+            Self::Wlr(obj) => obj.id(),
+            Self::Ext(obj) => obj.id(),
+        }
+    }
+
+    fn version(&self) -> u32 {
+        match self {
+            Self::Wlr(obj) => obj.version(),
+            Self::Ext(obj) => obj.version(),
+        }
+    }
+}
```

Comment:
I'd probably move it somewhere else to data_control only related file. Though, as alternative, you could write it with `WlrDataControl` and `ExtDataControl` as top-level types, would be much easier on a review, since the code will be pretty much duplicated.

And in fact, it's already pretty much duplicated, so really seems like there's no point in having a wrapper? Though, I could be wrong.

Reply:
Wrappers removed.